### ISSUE TITLE
Different zones in History have same color

### ISSUE THREAD

**Issue by ildar170975** [home-assistant/frontend#14705]
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.

### Describe the issue you are experiencing

Earlier different zones in History have different colors when displaying a history for `person` or `device_tracker` entities.
Since 2022.12 all zones have SAME color which makes graphs not easy to analyse.

### Describe the behavior you expected

1. Different zones must have different colors.
2. Same zone must have same color on graph.
3. If possible - same zone must have same color on all graphs.

### Steps to reproduce the issue

as above

### What version of Home Assistant Core has the issue?

2022.12.1

### In which browser are you experiencing the issue with?

Chrome 108.0.5359.99

**Comment by ildar170975** [comment]
Let me demonstrate an importance of this issue.

In an ideal scenario a `person` entity graph looks like:
![image](https://user-images.githubusercontent.com/71872483/207198078-0289733a-2d9b-4780-8ae1-7801ba65fd1c.png)
where a length of each part corresponds to a duration of the person's presence at Home or in some Zone.

But in a real life:
- `device_tracker` may not be updated in time -> a length of some part is longer / shorter than an actual duration;
- even some part may be missed - like "Home -> Zone 1, w/o Away":

![image](https://user-images.githubusercontent.com/71872483/207198440-acf1c79a-51c8-4ab8-aafb-8492cac7ed0e.png)

Also a zone's name may not be displayed if this part is short:
![image](https://user-images.githubusercontent.com/71872483/207198633-bc20bbd2-793e-4ff2-b3b8-1d77ad933256.png)

Imagine that a user wants to estimate where his child was today.
Surely a Logbook card with displaying time may be used etc.
But also this graph may be used:
![image](https://user-images.githubusercontent.com/71872483/207199005-6b5dc649-4cdd-41db-9faa-5e9fc50caf5b.png)
where the orange part is **clearly seen as a separate zone**.
Then the user may hover a mouse, see that this is a "Zone 1" in a baloon etc.

**Comment by midcode** [comment]
An even better solution would be if you could define specific colors for zones (like the icon you can already define)

**Comment by ildar170975** [comment]
@midcode 
Like that:
if zone = "church" -> `--history-zone-church-color`
if zone = "mom" -> `--history-zone-mom-color`
...

**Comment by prof7bit** [comment]
+1 for the above suggestion.

The frontend code already does dynamically construct variable names from entity attributes, you can find code like this 
````
--rgb-state-${domainColor}-color
```` 
for constructing the appropriate variable name for a certain domain.  Now you can easily bring this to the next level and also construct names like this 
````
--rgb-state-${domainColor}-$(entityId)-color
````
with even higher precedence, look up whether they are defined and if it is undefined then fall back to the more generic color for all entities of that domain.

This would than also allow stuff like
``` 
--rgb-state-binary-sensor-garagedoor-color: 'red'
```

And as a bonus, to make all the complaints about wrong colors disappear, make it so that
* if the color for a specific entity is undefined then fall back to the domain color
* if the domain color is undefined then fall back to the theme accent color
* remove many of the new default colors (so it will always fall back to active=yellow from the old theme everywhere) and let users color whatever they want/need **IF** they want/need.

**Comment by alexeyzel** [comment]
I have the same problem in version 2022.12.4 with binary sensors in the History Graph Card. Both zones (on and off) have the same colour: 
![image](https://user-images.githubusercontent.com/50808386/207436397-070310e0-d74c-4eb3-bbdc-17084c8f3440.png)
The same behaviour is on history tab too.
![image](https://user-images.githubusercontent.com/50808386/207436871-17970130-f2a8-41b7-8617-a21b81a2267b.png)

**Comment by alexeyzel** [comment]
Just updated to 2022.12.5 and my problem with binary sensors in the History Graph Card is fixed.

**Comment by ildar170975** [comment]
@alexeyzel 
Your issue is not related to Zones, it is about "for binary_sensors, same colors are used for on & off", and this is supposed to be fixed in 2022.12.5.

Ваше замечание не относится к теме данного тикета про раскраску Зон.
Ваше замечание касается бага "для binary_sensor используются одинаковые цвета для on & off в History" (можно найти соотв. тикет среди закрытых), данный баг должен был быть исправлен в 2022.12.5.

**Comment by ApriliaEdd** [comment]
> An even better solution would be if you could define specific colors for zones (like the icon you can already define)

This would be a nice solution if the person entity cards and badges also use this colour.  for example my wife works at multiple different sites i could make all these zones purple so easily see at a glance she is at work by the colour of the badge and the location text of the badge would show which site she was working at.

**Comment by ildar170975** [comment]
Selecting a color for icons, buttons and badges dependingly on a current zone is already possible: 
1. In a custom theme create your own variables:
```
user-zone-church-color: green
user-zone-school-color: red
```
2. For an icon, badge or button - use card-mod to define a color like:
```
color: var(—user-zone-{{states(config.entity)}}-color); 
```
This code is untested, just to explain the idea. 

But this approach is not possible now for History.
That is why it was suggested to add a functionality to draw a History using individual variables for each zone.
Not to mention a fact that many users do not use card-mod and expect to color icons, buttons and badges automatically.

**Comment by ApriliaEdd** [comment]
> Home Assistant must work without theme. Adding variables is not a solution for many users. We can not say to the user : "add this variable to your theme to fix the history issue".

I agree entirely with this, it would be a nice solution to select a colour for the zone when you create it just as you choose the icon ui and then have that colour match the history graph.   I know nothing of programming so this may or may not be achievable now or in the future

**Comment by prof7bit** [comment]
> Home Assistant must work without theme.

These variables would be optional: If it doesn't find a variable that explicitly defines the color for zone 'foo' it would then use the mechanism that colors zones in some automatic or default way. The same could be done even for individual entities. 

With a consistent and predictable naming scheme it would be possible (and relatively easy) for advanced users to add variables for coloring individual entities, individual zones, etc. and when the user does not define such variables it would just do it automatically in some reasonable way.

**Comment by ildar170975** [comment]
> Home Assistant must work without theme. Adding variables is not a solution for many users. We can not say to the user : "add this variable to your theme to fix the history issue".

Nobody advised here to use a custom theme as proper solution.
It was advised to provide automatically variables for each zone - and then users may re-define these variables as they like in their custom themes.

**Comment by ildar170975** [comment]
> i currently use card mod to achieve this

@ApriliaEdd 
In part of styling icons, buttons, badges etc - but not for History, I think.

**Comment by TroySK** [comment]
> Home Assistant must work without theme. Adding variables is not a solution for many users. We can not say to the user : "add this variable to your theme to fix the history issue".

@piitaya 
Unfortunately, this statement is unsupported by your commits and changes. The changes broke Home Assistant UIs for me and many users based on the number of issues on this repo. Instead of reverting the changes, you and the team are pushing even poor regression changes. This is not how you run a project being used by thousands of users.

This is a reflection of how poorly this project is being run. The breaking changes wasn't even mentioned in the release notes. It has been 15 days where the users have been struggling. 100s of issues, 1000s of comments and still no clue as to how do I fix my drunk Home Assistant.

A proprietary service cares more about its end users than this. Please revert all changes post 2022.12. Actions speak louder than words.

**Comment by marcelodavanzo** [comment]
Just came across this issue and it still exists in 2023.2. History graph for device_tracker is made illegible due to the same color being applied to all non home/away zones. 

The previous behaviour of random colors for individual zones was a much better approach and made the history graph actually meaningful at a glance given that the purpose of the frontend is to present data in a visually useful manner.

I'm not sure what coloring all non home/away zones the same is achieving but it certainly isn't making the graph any better than it was.

Would really like to see the previous behaviour back even if as a prop when creating device_trackers. I think providing customizeable variables for each zone is a bit overkill and would rather have HA handle this out of the box as piitaya said

**Comment by prof7bit** [comment]
It seems the different theme variables for different zones are working now. I found this out by reading through some very long threads in the forum and doing some experimentation on my own.

It would have been nice if he devs who are assigned to this (and similar) issues would have commented about the current state of affairs instead of just silently implementing something, leaving the issues open and uncommented and and not telling anybody anything about it.

**Comment by piitaya** [comment]
Hello 👋

One effect of the [new state colors](https://www.home-assistant.io/integrations/frontend/#state-color) introduced with `2023.2` is that you are able to customize the color for each zone.

For example you can define `state-person-my_office-color` or `state-device_tracker-my_office-color` if you have a zone named `My Office`. You can noticed that the zone name is slugified in the css variable (lower case and space replaced by `_`.
There is currently no way to customize it from the UI.

**Comment by ildar170975** [comment]
A possibility to define a custom color for a zone in a theme is a useful addition.
Thanks a lot for this.
But in a `history-graph` each zone should be colored differently BY DEFAULT.

Imagine a user who never dealt (or does not want to deal) with custom themes.
Or a user with 20 zones.

So, an ideal solution could be:
1. Use different colors by default for different zones:
- `zone.home` has SAME color in every graph;
- `zone.xyz` may have different colors on different graphs - but same color in a graph.
2. A user may define special colors for some zones in a custom theme (as it is already done).

**Comment by piitaya** [comment]
> If a user lives in not-english country - then probably he will have a `zone.parents` with a name in a local language ("Eltern"). Then these user-defined theme variables should be called `state-person-eltern-color`, right? And what about Korean language? Cyrillic letters? I do not think that non-latin letters may be used in variables.

We didn't added the state color variables to fix this issue. 

Person entities uses the friendly name as a state, not the entity_id so the state will be used. For information, we used these function to slugify the state : https://gist.github.com/hagemann/382adfc57adbd5af078dc93feef01fe1

> A possibility to define a custom color for a zone in a theme is a useful addition. Thanks a lot for this (although it may work in latin-letters countries only). But in a `history-graph` each zone should be colored differently BY DEFAULT.
> 
> Imagine a user who never dealt (or does not want to deal) with custom themes. Or a user with 20 zones.
> 
> So, an ideal solution could be:
> 
> 1. Use different colors by default for different zones:
> 
> * `zone.home` has SAME color in every graph;
> * `zone.xyz` may have different colors on different graphs - but same color in a graph.
> 
> 2. A user may define special colors for some zones in a custom theme.

I really think we need to better solution for this issue. I only provide the current state as @prof7bit asked an update. For now, I don't have a perfect solution for this.

**Comment by ildar170975** [comment]
> We didn't added the state color variables to fix this issue.

Correct, and I wrote about user-defined variables based on [this approach](https://www.home-assistant.io/integrations/frontend/#state-color).

> Person entities uses the friendly name as a state, not the entity_id so the state will be used

This is exactly what I wrote. And if friendly names are in Korean or Cyrillic (for instance) - then [that approach](https://www.home-assistant.io/integrations/frontend/#state-color) cannot be used.

**Comment by Mariusthvdb** [comment]
> `state-person-my_office-color` or `state-device_tracker-my_office-color`

would there be an easy way out of this scenario?:

home: green
not_home: gray (same as not in a zone, and not home= travelling)
in a zone (have a lot of zones, so a generic 'zone' color would be nice): orange

ideally this would  be in core somehow (might already be, so please forgive me if the case), or be added to it as minimum requirement?

**Comment by ildar170975** [comment]
Up

**Comment by ildar170975** [comment]
up

**Comment by weber8thomas** [comment]
up

**Comment by Chacsam** [comment]
up

**Comment by ildar170975** [comment]
up

**Comment by nzrutman** [comment]
Perhaps we can get some movement on this if we take it back down to the original request - I don't need custom colors or GUI blah, I just need different zones to be different colors in device_tracker history. It SUCKS that all zones except for home are blue - completely unreadable to see history state changes, which is the whole point of a state history graph.

![Screenshot 2024-09-24 at 11 27 47 AM](https://github.com/user-attachments/assets/b47eec52-3eb7-493f-9913-ad8f8a03c08a)
When did these change state? No idea. Were there any short transitional locations between them?  No idea.

**Comment by eb3nezer** [comment]
Yes still seeing this.

**Comment by jlm70** [comment]
Same desire here... when you setup many zones, the graph becomes somewhat unuseful.
It would be great to connect colors to zones, such as for the icons. Thanks!

**Comment by jacATK** [comment]
I would also like to assign different colors to different zones. I have eight zones defined and the only one not blue is my home zone.

**Comment by ildar170975** [comment]
Up

**Comment by ildar170975** [comment]
Hopefully can be solved with https://github.com/home-assistant/frontend/pull/25647

**Comment by Chacsam** [comment]
Still no improvement on 2025.11.2
Home is green;
Away is grey;
everything else is blue

**Comment by ShadowJonathan** [comment]
I'd say #29700 does not fix this issue? It allows you to define the color for a slugified location, but it doesn't change the default of blue-on-blue :/

**Comment by MindFreeze** [comment]
> but it doesn't change the default of blue-on-blue :/

It does

**Comment by ShadowJonathan** [comment]
```ts
  // otherwise return undefined to get unique colors from the generic color handler.
```

i didn't properly read the comment, as this behaviour was a bit too implicit/magic, but alright then

**Comment by ShadowJonathan** [comment]
I can confirm this works now, thanks!!
